### PR TITLE
Fix huber loss

### PR DESCRIPTION
**Pull request by henry0312**
TODO:

- [x] merge and rebase #175

### CODE REVIEW

File: src/objective/objective_function.cpp [hunk 1]
```diff
@@ -8,9 +8,14 @@ namespace LightGBM {
 
 ObjectiveFunction* ObjectiveFunction::CreateObjectiveFunction(const std::string& type, const ObjectiveConfig& config) {
   if (type == std::string("regression")) {
+    // deprecated
     return new RegressionL2loss(config);
+  } else if (type == std::string("regression_l2") || type == std::string("mean_squared_error") || type == std::string("mse")) {
+    return new RegressionL2loss(config);
```
Comment (guolinke) [inline src/objective/objective_function.cpp]: merge these two if?

File: src/objective/regression_objective.hpp [hunk 2]
```diff
@@ -51,13 +51,88 @@ class RegressionL2loss: public ObjectiveFunction {
...
+
+        /*
+         * approximate hessians with Gaussian function
+         * cf. https://en.wikipedia.org/wiki/Gaussian_function
+         */
+        const double x = (std::fabs(diff) > 0.0) ? std::fabs(diff) : 1.0e-6;
+        const double a = 2.0 * weights_[i];  // difference of two first derivatives, (zero to inf) and (zero to -inf).
+        const double b = 0.0;
+        const double c = (std::fabs(score[i]) + std::fabs(label_[i])) / 1.0e3;
+        hessians[i] = std::exp(-(x - b) * (x - b) / 2.0 * c * c) / a;
```
Comment (guolinke) [inline src/objective/regression_objective.hpp]: I think you can use an inline function to calculate hessian and avoid these repeated code.

File: src/objective/regression_objective.hpp [hunk 3]
```diff
@@ -81,13 +157,23 @@ class RegressionLHuberLoss: public ObjectiveFunction {
...
+
+            /*
+             * approximate hessians with Gaussian function
+             * cf. https://en.wikipedia.org/wiki/Gaussian_function
+             */
+            const double x = (std::fabs(diff) > 0.0) ? std::fabs(diff) : 1.0e-6;
+            const double a = 2.0;  // difference of two first derivatives, (zero to inf) and (zero to -inf)
+            const double b = 0.0;
+            const double c = (std::fabs(score[i]) + std::fabs(label_[i])) / 1.0e3;
+            hessians[i] = std::exp(-(x - b) * (x - b) / 2.0 * c * c) / a;
```
Comment (guolinke) [inline src/objective/regression_objective.hpp]: same as above, and you can put this function into utils/common.h

File: src/objective/objective_function.cpp (hunk 1 above)
Comment (henry0312) [inline src/objective/objective_function.cpp]: I'm sorry, I don't understand what you mean 😢
This PR will be merged after #175

File: src/objective/regression_objective.hpp (hunk 3 above)
Comment (henry0312) [inline src/objective/regression_objective.hpp]: ok!

File: src/objective/objective_function.cpp (hunk 1 above)
Comment (guolinke) [inline src/objective/objective_function.cpp]: change 
```
if () 
...
else if()
...
```
to 
```
if()
...
```
. They are all regression objective, right?

Comment (henry0312) [inline src/objective/objective_function.cpp]: yes, they are.

you mean, 

```cpp
  if (type == std::string("regression") || type == std::string("regression_l2") || type == std::string("mean_squared_error") || type == std::string("mse")) {
    return new RegressionL2loss(config);
  }
```

right?

Comment (guolinke) [inline src/objective/objective_function.cpp]: yes. and you can break ```if()``` to two line:
```
  if (type == std::string("regression") || type == std::string("regression_l2") 
      || type == std::string("mean_squared_error") || type == std::string("mse")) {
    return new RegressionL2loss(config);
  }
```

**Comment by henry0312** [comment]
rebased master.
However, there may be something wrong with `ApproximateHessianWithGaussian`.
I'll check.

**Comment by Laurae2** [comment]
@henry0312 This [topic](https://www.kaggle.com/c/allstate-claims-severity/forums/t/24520/effect-of-mae) has a bunch of Gradient/Hessian functions for helping to converge for a L1 loss function when a hessian is mandatory for convergence.

They were all tested in xgboost and are all converging fast. Without a hessian, the convergence is slow (which seems to be the case in LightGBM also as @guolinke mentioned in the previous PR for L1 loss). It would be good to have an example to check the convergence speed of the L1 loss.

The short "explanation" is that when a hessian is mandatory for good convergence and we want to use a L1 loss function, we should avoid at any cost a constant hessian. To fix this issue, we smooth the linear function using a curve so the hessian is never zeroed out (better and faster convergence speed). For instance, the [Fair objective](http://research.microsoft.com/en-us/um/people/zhang/INRIA/Publis/Tutorial-Estim/node24.html) (from a Microsoft research) is used to approximate such L1 function requiring a hessian.

The R versions with their gradient/hessian formulas are below:

```r
ln_cosh_obj <- function(preds, dtrain) {
  labels <- getinfo(dtrain, "label")
  grad <- tanh(preds-labels)
  hess <- 1-grad*grad
  return(list(grad = grad, hess = hess))
}

ln_expexp_obj <- function(preds, dtrain) {
  labels <- getinfo(dtrain, "label")
  x <- preds-labels
  grad <-  (exp(2*x)-1) / (exp(2*x)+1)
  hess <-  (4*exp(2*x)) / (exp(2*x) + 1)^2 
  return(list(grad = grad, hess = hess))
}

cauchyobj <- function(preds, dtrain) {
  labels <- getinfo(dtrain, "label")
  c <- 2  #the lower the "slower/smoother" the loss is
  x <-  preds-labels
  grad <- x / (x^2/c^2+1)
  hess <- -c^2*(x^2-c^2)/(x^2+c^2)^2
  return(list(grad = grad, hess = hess))
}

fairobj <- function(preds, dtrain) {
  labels <- getinfo(dtrain, "label")
  c <- 2  #the lower the "slower/smoother" the loss is
  x <-  preds-labels
  grad <- c*x / (abs(x)+c)
  hess <- c^2 / (abs(x)+c)^2
  return(list(grad = grad, hess = hess))
}
```

**Comment by henry0312** [comment]
@Laurae2 the hessian of mae is delta function, therefore, if we can approximate it with something appropriate (e.g. gaussian function), converge and performance will be good, I guess.

File: include/LightGBM/utils/common.h [hunk 4]
```diff
@@ -405,17 +405,18 @@ inline void SortForPair(std::vector<T1>& keys, std::vector<T2>& values, size_t s
...
 * y is a prediction.
-* t mesas true target.
+* t means true target.
+* g means gradient.
 * w means weights.
 */
-inline static double ApproximateHessianWithGaussian(double y, double t, double w=1.0f) {
+inline static double ApproximateHessianWithGaussian(const double y, const double t, const double g, const double w=1.0f) {
   const double diff = y - t;
   const double pi = M_PI;
```
Comment (guolinke) [inline include/LightGBM/utils/common.h]: where is the define of ```M_PI``` ? I cannot compile it on windows.

Comment (henry0312) [inline include/LightGBM/utils/common.h]: Could you add `#include<cmath>`?

Comment (guolinke) [inline include/LightGBM/utils/common.h]: it seems need add ```#define _USE_MATH_DEFINES``` in windows:
http://stackoverflow.com/questions/6563810/m-pi-works-with-math-h-but-not-with-cmath-in-visual-studio

Comment (henry0312) [inline include/LightGBM/utils/common.h]: probably fixed with https://github.com/Microsoft/LightGBM/pull/178/commits/05e97f10cb2b55774ff16164b3676da5cb7dc731

**Comment by henry0312** [comment]
I finished fixing `ApproximateHessianWithGaussian` (also Huber loss).
L1 loss is approximated with Gaussian function like http://www.wolframalpha.com/input/?i=Integrate%5BErf%5Bx%5D%5D,+Integrate%5BErf%5Bx%2F10%5D%5D.

**Comment by henry0312** [comment]
Next, I will create a Fair loss in an another PR, which @Laurae2 tells me.
